src/Arcus.WebApi.Security/Authentication/CertificateAuthenticationConfig.cs

```diff
@@ -0,0 +1,172 @@
+        {
+            Guard.NotNull(location, nameof(location), "Location implementation to retrieve the expected thumbprint cannot be 'null'");
+            Guard.NotNullOrWhitespace(configuredKey, nameof(configuredKey), "Configured key to retrieve expected thumbprint cannot be blank");
+
+            return AddCertificateRequirement(X509ValidationRequirement.Thumbprint, location, configuredKey);
+        }
+
+        private CertificateAuthenticationConfig AddCertificateRequirement(
+            X509ValidationRequirement requirement,
+            IX509ValidationLocation location,
+            string configuredKey)
+        {
+            Guard.NotNull(location, nameof(location), "Location cannot be 'null'");
+            Guard.NotNullOrWhitespace(configuredKey, nameof(configuredKey), "Configured key cannot be blank");
+
+            // Overwrites existing requirements.
+            _locationAndKeyByRequirement[requirement] = (location, new ConfiguredKey(configuredKey));
+            
+            return this;
+        }
+
+        private static IX509ValidationLocation GetValidationLocationImplementation(X509ValidationLocation location)
+        {
+            switch (location)
+            {
+                case X509ValidationLocation.SecretProvider:
+                    return SecretProviderValidationLocation.Instance;
+                case X509ValidationLocation.Configuration:
+                    return ConfigurationValidationLocation.Instance;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(location), location, "Unknown certificate validation location");
+            }
+        }
+
+        /// <summary>
+        /// Gets all the expected <see cref="X509Certificate2"/> values from this current configuration instance.
+        /// </summary>
+        /// <param name="services">The request services to retrieve the necessary implementations during the retrieval of each expected certificate value.</param>
+        /// <param name="logger">The logger used during the retrieval of each expected certificate value.</param>
+        /// <returns>The key/value pair of which certificate requirement to validate together with which value in the actual client certificate to expect.</returns>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="services"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="logger"/> is <c>null</c>.</exception>
+        internal async Task<IDictionary<X509ValidationRequirement, ExpectedCertificateValue>> GetAllExpectedCertificateValues(IServiceProvider services, ILogger logger)
+        {
+            Guard.NotNull(services, nameof(services), "Request services cannot be 'null'");
+            Guard.NotNull(logger, nameof(logger), "Logger cannot be 'null'");
+
+            var expectedValuesByRequirement = await Task.WhenAll(_locationAndKeyByRequirement.Select(async keyValue =>
```

Comment:
This is another case where all the logic in the select should be moved to a method instead

Reply:
Thanks for moving @stijnmoreels 	!